src/extensions/Statiq.ApiClient/ReadApiWithClient.cs

```diff
@@ -0,0 +1,149 @@
+            return this;
+        }
+
+        /// <summary>
+        /// Initialize the API client.
+        /// </summary>
+        /// <param name="init">An API client initialization action.</param>
+        /// <returns>The current module instance.</returns>
+        public ReadApiWithClient<TClient> WithClientInitialization(Action<IDocument, IExecutionContext, TClient> init)
+        {
+            if (init == null)
+            {
+                throw new ArgumentException("Argument is null or empty", nameof(init));
+            }
+
+            _init = init.ThrowIfNull(nameof(init));
+            return this;
+        }
+
+        /// <summary>
+        /// Submits the API client name (for logging).
+        /// </summary>
+        /// <param name="name">Client name.</param>
+        /// <returns>The current module instance.</returns>
+        public ReadApiWithClient<TClient> WithClientName(string name)
+        {
+            _clientName = name.ThrowIfNullOrWhiteSpace(nameof(name));
+            return this;
+        }
+
+        /// <summary>
+        /// Submits a request to the API client. This allows you to incorporate data from the execution context in your request.
+        /// </summary>
+        /// <param name="key">The metadata key in which to store the return value of the request function.</param>
+        /// <param name="request">A function with the request to make.</param>
+        /// <returns>The current module instance.</returns>
+        public ReadApiWithClient<TClient> WithRequest(string key, Func<IExecutionContext, TClient, object> request)
+        {
+            key.ThrowIfNullOrWhiteSpace(nameof(key));
+            request.ThrowIfNull(nameof(request));
+
+            _requests[key] = (doc, ctx, client) => request(ctx, client);
+            return this;
+        }
+
+        /// <summary>
+        /// Submits a request to the API. This allows you to incorporate data from the execution context and current document in your request.
+        /// </summary>
+        /// <param name="key">The metadata key in which to store the return value of the request function.</param>
+        /// <param name="request">A function with the request to make.</param>
+        /// <returns>The current module instance.</returns>
+        public ReadApiWithClient<TClient> WithRequest(string key, Func<IDocument, IExecutionContext, TClient, object> request)
+        {
+            key.ThrowIfNullOrWhiteSpace(nameof(key));
+
+            _requests[key] = request.ThrowIfNull(nameof(request));
+            return this;
+        }
+
+        /// <inheritdoc/>
+        protected override IEnumerable<IDocument> ExecuteInput(IDocument input, IExecutionContext context)
+        {
+            _init?.Invoke(input, context, _client);
+            ConcurrentDictionary<string, object> results = new ConcurrentDictionary<string, object>();
+            System.Threading.Tasks.Parallel.ForEach(_requests, request =>
```

Comment:
I wonder if executing the client in parallel should be an option, or maybe just do it serially to be safe. Some APIs don't take nicely to parallel requests due to rate limits, etc.

Reply:
Maybe should use:

```csharp

private int _maxDegreeOfParallelism = -1;

...

/// <summary>
/// Submits the maximum number of enabled concurrent tasks for executing requests.
/// </summary>
/// <param name="maxDegreeOfParallelism">The maximum number of enabled concurrent tasks for executing requests. Put 1 to execute requests synchronously.</param>
/// <returns>The current module instance.</returns>
public ReadApi<TClient> WithMaxDegreeOfParallelism(int maxDegreeOfParallelism)
{
    _maxDegreeOfParallelism = maxDegreeOfParallelism;
    return this;
}

...

protected override Task<IEnumerable<IDocument>> ExecuteInputAsync(IDocument input, IExecutionContext context)
{
    ...
    System.Threading.Tasks.Parallel.ForEach(_requests, new ParallelOptions { MaxDegreeOfParallelism = _maxDegreeOfParallelism }, request => { ... }
    ...
}
```

What do you think?

Reply:
I think making the throttling configurable is on the right track. Thought about this a little more and realized the underlying problem I was trying to address is rate limits specifically, the parallelism doesn't really matter besides being a function of that. I.e., I do a lot of GitHub API work and can see using this module for that, but the GitHub API has strictly controlled rate limits. Even limiting to a single concurrent operation won't totally solve that problem.

I wonder if an approach similar to https://codeburst.io/throttling-concurrent-outgoing-http-requests-in-net-core-404b5acd987b would work (not so much the `HttpClient` bits, but the strategy of using a `SemaphoreSlim` with task waits)? Then we would provide both a `WithRequestLimit()` setting and a `WithRequestDelay()` setting. What do you think of that idea?

Reply:
This possibility should be explored. I'll get back a little later.

Reply:
@daveaglick done! ready for review.